File: group_assignment/concave.py
import pygame
from OpenGL.GL import *
from OpenGL.GLUT import *
from OpenGL.GLU import *
import numpy as np

# Point structure to store vertices
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

# Original concave polygon (arrowhead shape)
concavePoly = [
    Point(100, 200),  # P1
    Point(150, 250),  # P2
    Point(200, 200),  # P3
    Point(150, 150),  # P4 <- inward (concave)
    Point(150, 100)   # P5
]

# Cross product (z-component only)
def crossZ(a, b, c):
    v1x = b.x - a.x
    v1y = b.y - a.y
    v2x = c.x - b.x
    v2y = c.y - b.y
    return v1x * v2y - v1y * v2x

# Find concave vertex index
def findConcaveIndex(poly):
    n = len(poly)
    for i in range(n):
        a = poly[(i - 1 + n) % n]
        b = poly[i]
        c = poly[(i + 1) % n]
        if crossZ(a, b, c) < 0:
            return i
    return -1

# Function to draw a polygon
def drawPolygon(poly, r, g, b):
    glColor3f(r, g, b)
    glBegin(GL_POLYGON)
    for p in poly:
        glVertex2f(p.x, p.y)
    glEnd()

# Function to draw text (a label)
def drawText(x, y, label):
    glRasterPos2f(x, y)
    for c in label:
        glutBitmapCharacter(GLUT_BITMAP_HELVETICA_18, ord(c))

# Function to draw a line
def drawLine(a, b, r=0, g=0, bCol=0):
    glColor3f(r, g, bCol)
    glBegin(GL_LINES)
    glVertex2f(a.x, a.y)
    glVertex2f(b.x, b.y)
    glEnd()

# The main display function
def display():
    glClear(GL_COLOR_BUFFER_BIT)

    # Draw the original polygon (left side)
    glColor3f(1.0, 0.0, 0.0)  # Red outline
    glBegin(GL_LINE_LOOP)
    for p in concavePoly:
        glVertex2f(p.x, p.y)
    glEnd()
    drawText(90, 270, "Original Concave Polygon")

    # Detect the concave vertex
    concaveIndex = findConcaveIndex(concavePoly)
    if concaveIndex == -1:
        return

    n = len(concavePoly)

    # Pick a valid target vertex to split with
    splitTarget = 1
    if splitTarget == concaveIndex or (splitTarget + 1) % n == concaveIndex or (splitTarget - 1 + n) % n == concaveIndex:
        splitTarget = 4  # Try P5 if P2 is adjacent

    # Offset polygon to the right (draw it on the right side of the window)
    offsetPoly = [Point(p.x + 300, p.y) for p in concavePoly]

    # Create two split polygons
    poly1 = []
    poly2 = []

    # Build the first polygon (clockwise)
    i = concaveIndex
    while True:
        poly1.append(offsetPoly[i])
        if i == splitTarget:
            break
        i = (i + 1) % n

    # Build the second polygon (counter-clockwise)
    i = concaveIndex
    while True:
        poly2.append(offsetPoly[i])
        if i == (splitTarget - 1 + n) % n:
            break
        i = (i - 1 + n) % n
    poly2.append(offsetPoly[splitTarget])  # Complete the second polygon

    # Draw the split polygons (right side)
    drawPolygon(poly1, 0.0, 1.0, 0.0)  # Green polygon
    drawPolygon(poly2, 0.0, 0.0, 1.0)  # Blue polygon

    # Draw the diagonal used for splitting
    drawLine(concavePoly[concaveIndex], concavePoly[splitTarget], 0, 0, 0)  # Left
    drawLine(offsetPoly[concaveIndex], offsetPoly[splitTarget], 0, 0, 0)   # Right

    glColor3f(0, 0, 0)  # Black outline for the offset polygon
    glBegin(GL_LINE_LOOP)
    for p in offsetPoly:
        glVertex2f(p.x, p.y)
    glEnd()

    drawText(360, 270, "Split Using Vector Method")

    pygame.display.flip()

# Initialize pygame and set up OpenGL
def init():
    pygame.init()
    display_width, display_height = 700, 300
    pygame.display.set_mode((display_width, display_height), pygame.DOUBLEBUF | pygame.OPENGL)
    gluOrtho2D(0, display_width, 0, display_height)

# Main function to start everything
def main():
    init()
    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                return

        display()

# Run the program
if __name__ == "__main__":
    main()

File: group_assignment/concave.cpp
#include <GL/glut.h>
#include <iostream>
#include <vector>
#include <cmath>

struct Point
{
    float x, y;
    Point(float x_, float y_) : x(x_), y(y_) {}
};

std::vector<Point> concavePoly = {
    Point(100, 200), // P1
    Point(150, 250), // P2
    Point(200, 200), // P3
    Point(150, 150), // P4 <- concave
    Point(150, 100)  // P5
};

float crossZ(const Point &a, const Point &b, const Point &c)
{
    float abx = b.x - a.x;
    float aby = b.y - a.y;
    float bcx = c.x - b.x;
    float bcy = c.y - b.y;
    return abx * bcy - aby * bcx;
}

int findConcaveIndex(const std::vector<Point> &poly)
{
    int n = poly.size();
    for (int i = 0; i < n; ++i)
    {
        Point prev = poly[(i - 1 + n) % n];
        Point curr = poly[i];
        Point next = poly[(i + 1) % n];
        if (crossZ(prev, curr, next) < 0)
        {
            return i;
        }
    }
    return -1;
}

Point vectorIntersection(const Point &p1, const Point &p2, const Point &q1, const Point &q2)
{
    auto det = [](float a, float b, float c, float d)
    {
        return a * d - b * c;
    };

    float a1 = p2.y - p1.y, b1 = p1.x - p2.x;
    float c1 = a1 * p1.x + b1 * p1.y;

    float a2 = q2.y - q1.y, b2 = q1.x - q2.x;
    float c2 = a2 * q1.x + b2 * q1.y;

    float determinant = det(a1, b1, a2, b2);
    if (determinant == 0)
        return Point(-1, -1); // Parallel lines

    float x = det(c1, b1, c2, b2) / determinant;
    float y = det(a1, c1, a2, c2) / determinant;
    return Point(x, y);
}

void drawPolygon(const std::vector<Point> &poly, const float color[3])
{
    glColor3fv(color);
    glBegin(GL_POLYGON);
    for (const auto &p : poly)
    {
        glVertex2f(p.x, p.y);
    }
    glEnd();
}

void drawOutline(const std::vector<Point> &poly, const float color[3])
{
    glColor3fv(color);
    glBegin(GL_LINE_LOOP);
    for (const auto &p : poly)
    {
        glVertex2f(p.x, p.y);
    }
    glEnd();
}

void drawLine(const Point &a, const Point &b, const float color[3])
{
    glColor3fv(color);
    glBegin(GL_LINES);
    glVertex2f(a.x, a.y);
    glVertex2f(b.x, b.y);
    glEnd();
}

std::vector<Point> translatePolygon(const std::vector<Point> &poly, float dx, float dy)
{
    std::vector<Point> translated;
    for (const auto &p : poly)
    {
        translated.push_back(Point(p.x + dx, p.y + dy));
    }
    return translated;
}

void splitPolygon(const std::vector<Point> &poly, std::vector<Point> &poly1, std::vector<Point> &poly2, int &concaveIndex, int &splitIndex)
{
    int n = poly.size();
    concaveIndex = findConcaveIndex(poly);

    if (concaveIndex == -1)
    {
        poly1 = poly; // Already convex
        return;
    }

    // Find a valid split line using vector intersection
    Point concavePoint = poly[concaveIndex];
    for (int i = 0; i < n; ++i)
    {
        if (i == concaveIndex || i == (concaveIndex - 1 + n) % n || i == (concaveIndex + 1) % n)
        {
            continue;
        }
        Point splitPoint = poly[i];

        // Check if the line intersects with any other edges
        bool valid = true;
        for (int j = 0; j < n; ++j)
        {
            Point p1 = poly[j];
            Point p2 = poly[(j + 1) % n];
            if (p1.x == concavePoint.x && p1.y == concavePoint.y || p2.x == concavePoint.x && p2.y == concavePoint.y ||
                p1.x == splitPoint.x && p1.y == splitPoint.y || p2.x == splitPoint.x && p2.y == splitPoint.y)
            {
                continue;
            }
            if (vectorIntersection(concavePoint, splitPoint, p1, p2).x != -1)
            {
                valid = false;
                break;
            }
        }

        if (valid)
        {
            splitIndex = i;
            break;
        }
    }

    // Form two new polygons
    poly1.clear();
    int i = concaveIndex;
    while (true)
    {
        poly1.push_back(poly[i]);
        if (i == splitIndex)
        {
            break;
        }
        i = (i + 1) % n;
    }

    poly2.clear();
    i = concaveIndex;
    while (true)
    {
        poly2.push_back(poly[i]);
        if (i == (splitIndex - 1 + n) % n)
        {
            break;
        }
        i = (i - 1 + n) % n;
    }
    poly2.push_back(poly[splitIndex]);
}

void display()
{
    glClear(GL_COLOR_BUFFER_BIT);

    // Draw the original concave polygon on the left
    const float red[3] = {1.0f, 0.0f, 0.0f};
    drawOutline(concavePoly, red);

    // Split the polygon into two convex parts
    std::vector<Point> poly1, poly2;
    int concaveIndex, splitIndex;
    splitPolygon(concavePoly, poly1, poly2, concaveIndex, splitIndex);

    // Draw the split line on the original polygon
    const float black[3] = {0.0f, 0.0f, 0.0f};
    drawLine(concavePoly[concaveIndex], concavePoly[splitIndex], black);

    // Translate the split polygons to the right side with more spacing
    poly1 = translatePolygon(poly1, 350, 50);
    poly2 = translatePolygon(poly2, 350, -50);

    // Draw the split polygons on the right
    const float green[3] = {0.3f, 0.8f, 0.3f};
    drawPolygon(poly1, green);
    drawOutline(poly1, green);

    const float blue[3] = {0.2f, 0.4f, 0.9f};
    drawPolygon(poly2, blue);
    drawOutline(poly2, blue);

    glutSwapBuffers();
}

void initOpenGL()
{
    glClearColor(1.0f, 1.0f, 1.0f, 1.0f); // Set background color to white
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    glOrtho(0, 700, 0, 300, -1, 1); // Set orthographic view
}

int main(int argc, char **argv)
{
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(700, 300);
    glutCreateWindow("Concave Polygon Split");

    initOpenGL();

    glutDisplayFunc(display); // Set display function
    glutMainLoop();           // Enter GLUT event processing loop

    return 0;
}
